### PR TITLE
Add the collection object: A set of components

### PR DESCRIPTION
**Pull request by Hywan**
Related to https://github.com/fruux/sabre-vobject/issues/189.
It does not fix this issue but this is needed to solve it.

### CODE REVIEW

File: lib/Collection.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+     * @param  mixed  $offset    Component offset.
+     * @param  mixed  $value     Component.
+     * @return Collection
+     * @throw  \InvalidArgumentException
+     */
+    function offsetSet($offset, $value) {
+
+        $type = $this->type;
+
+        if (!(($value instanceof Document) && ($value instanceof $type))) {
```
Comment (staabm) [inline lib/Collection.php]: couldn't you instead check whether `$type` is `is_subclass_of Document` in `__construct` and check here only check `instanceof $type` then?

Comment (Hywan) [inline lib/Collection.php]: Unfortunately, we cannot have `$type instanceof Document` where `$type` is a string.

**Comment by Hywan** [comment]
The build fails on CS that are not related to this PR. These CS are, by the way, fixed in #160.

File: lib/Collection.php (hunk 1 above)
Comment (staabm) [inline lib/Collection.php]: therefore I suggest `is_subclass_of`

Comment (Hywan) [inline lib/Collection.php]: :+1:

**Comment by Hywan** [comment]
I wonder if we need to define `Collection` as a `Node`. It would help a lot. I met this usecase while developing xCard support. I am going to update this PR.

**Comment by Hywan** [comment]
This PR **must not** be merged before #160, because `Node` changes.

**Comment by Hywan** [comment]
Actually, extending `Node` will not help. We have to extend `Component` in order to work seamlessly with `Writer`. Hmm… I think it's better to keep `Collection` as it is and extend the `Writer` to deal with it.

**Comment by evert** [comment]
I see this is already closed, but I would wonder.. is this needed at all? The splitter classes exists because it allows someone to progressively parse many objects, without loading them all in memory. It looks like the intent of this class is to actually load everything in memory.

I feel that for many real-life use-cases, this would not be desirable.

One idea could be to merge the 'getNext()` method from the splitter right into the parser. That way you can start parsing ones, and if you know there may be more in a file, you can call it multiple times.

Regardless, since we can start depending on PHP 5.5, whenever you consider building an iterator, consider using a generator instead ;)

**Comment by Hywan** [comment]
@evert That was exactly my plan :-]. You read in my minds!

**Pull request closed without merging**